messaging/activity/FirebaseMessagingActivityGenerator.cs

```diff
@@ -0,0 +1,172 @@
+"  // ID. Used by upstream, and set by GmsCore.",
+"  private static final String EXTRA_MESSAGE_ID_KEY = \"google.message_id\";",
+"",
+"  // The key in the intent's extras that maps to the incoming message's sender value.",
+"  private static final String EXTRA_FROM = \"google.message_id\";",
+"",
+"  /**",
+"   * Workaround for when a message is sent containing both a Data and Notification payload.",
+"   *",
+"   * <p>When the app is in the background, if a message with both a data and notification payload is",
+"   * received the data payload is stored on the Intent passed to onNewIntent. By default, that",
+"   * intent does not get set as the Intent that started the app, so when the app comes back online",
+"   * it doesn't see a new FCM message to respond to. As a workaround, we override onNewIntent so",
+"   * that it sends the intent to the MessageForwardingService which forwards the message to the",
+"   * FirebaseMessagingService which in turn sends the message to the application.",
+"   */",
+"  @Override",
+"  protected void onNewIntent(Intent intent) {{",
+"    super.onNewIntent(intent);",
+"",
+"    // If we do not have a 'from' field this intent was not a message and should not be handled. It",
+"    // probably means this intent was fired by tapping on the app icon.",
+"    Bundle extras = intent.getExtras();",
+"    if (extras == null) {{",
+"      return;",
+"    }}",
+"    String from = extras.getString(EXTRA_FROM);",
+"    String messageId = extras.getString(EXTRA_MESSAGE_ID_KEY);",
+"    if (messageId == null) {{",
+"      messageId = extras.getString(EXTRA_MESSAGE_ID_KEY_SERVER);",
+"    }}",
+"    if (from != null && messageId != null) {{",
+"      Intent message = new Intent(this, MessageForwardingService.class);",
+"      message.setAction(MessageForwardingService.ACTION_REMOTE_INTENT);",
+"      message.putExtras(intent);",
+"      message.setData(intent.getData());",
+"      MessageForwardingService.enqueueWork(this, message);",
+"    }}",
+"    setIntent(intent);",
+"  }}",
+"",
+"  /**",
+"   * Dispose of the mUnityPlayer when restarting the app.",
+"   *",
+"   * <p>This ensures that when the app starts up again it does not start with stale data.",
+"   */",
+"  @Override",
+"  protected void onCreate(Bundle savedInstanceState) {{",
+"    if (mUnityPlayer != null) {{",
+"      mUnityPlayer.quit();",
+"      mUnityPlayer = null;",
+"    }}",
+"    super.onCreate(savedInstanceState);",
+"  }}",
+"}}"
+  };
+  private readonly string BaseActivityClass = "UnityPlayerActivity";
+  private readonly string BaseGameActivityClass = "UnityPlayerGameActivity";
+
+  private readonly string GeneratedFileTag = "FirebaseMessagingActivityGenerated";
+  // If this tag is present on the generated file, it will not be replaced.
+  private readonly string PreserveTag = "FirebasePreserve";
+
+  private readonly string OutputPath = Path.Combine("Plugins", "Android");
```

Comment:
I have a feeling you need to put Android related files under a folder name ended with `.androidlib`, ex. `Assets/Plugins/Android/FirebaseMessaging.androidlib`.

https://docs.unity3d.com/Manual/android-library-project-import.html

Reply:
I think that is only if you are using a prebuilt library, or need an Android Manifest file.  For pure java files, it just needs to be present and set to Android: https://docs.unity3d.com/Manual/android-java-and-kotlin-plugins-create.html

Reply:
The instruction mentioned that you need to enable `Android` in platform, which can be done in `.meta` metadata file.

Have you try to build the project and see if this `FirebaseMessagingActivityEditor.java` is included in the generated Gradle project? You can also use export project (a build option), which should save you time to find the generated Gradle project under `Library` folder.

Reply:
Yeah, from all my tests Unity detects it is a java file, and automatically sets it up to just be included in the Android build.  This worked for both building directly in Unity, and exporting to Gradle.